Feature/implement certification 4.7.9 ii

RanRegev opened the pull request:
This PR replaces PR #1114
This PR contains the entire agent flow to pass certification test 4.7.9 with ETH backhaul.
It added two tlvs (beacon query and response), it takes care for sending the request only to the correct radio, and it handles the response.
The result of the full flow is a beacon measurements response, 1905 format, received at the controller. The data from the message is NOT stored or processed in any way.

RanRegev committed:
auto generated: file mode was changed

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
tlv: added beacon query and response

added the yamls files to tlv_conf.py
fixed query yaml
added generated .h and .cpp

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
agent: slave: implemented handling 1905 beacon request

implementation:
wiring (making sure the 1905 message is handled)
converting 1905 to vendor specific message
forward the vs to the monitor

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
agent: slave: handle vs beacon response

flow: translates the vendor specific message
into a 1905 message and forward to the controller

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
controller: handling ack with error and receiving 1905 beacon response

Handling ack message with error-tlv
Receiving beacon response - doing nothing with it
except for logging.

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
framework: transport: improved warning log message

added detailes to a too generic warning message

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
tests: test_flow: added beacon query request test

The test only generates the request, no tests
are actually taken

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
agent: slave: fine tuning 1905->vs transmormation

It seems that the transformation that is done up
to this change is missing. Added operatrion class
and report.

This is probably the first change in a serious of changes
in this transformation:
Transform No. I

MAP-4.7.9_ETH:netgear-rax40

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
dwpal: monitor: added a todo comment

non functional, just a reminder that there is a work to be done

Signed-off-by: Ran regev <ran.regev@devalore.com>

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -503,11 +503,17 @@ bool monitor_thread::update_sta_stats()
+            }
             return false;
         }
+        stat_state_changed = true;
```
I think this is incorrect.
You prevent all error messages not just for the specific `sta_mac`.
Also, it's probably important to see in the logs that something doesn't work and we're unable to update the station statistics.

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -100,6 +101,35 @@ namespace beerocks {
+/////////////////////////////// Helper Functions /////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+template <class SlavesContainer, class MAC>
```
Any specific reason for templating this?

RanRegev replied:
The original thought was that we have mac as a struct and mac as a string, so by specializing the template with string it would be transparent to the user. This thought lead to another thought, since it is a helper function, loosely coupled to the backhaul (the return type is a backhaul inner type) why not free the container to be whatever it is, as long as it is forward-iterable.

arnout replied:
I don't like this either, for either template argument.
- MAC should always be sMacAddr. If there's a string anywhere, call mac_from_string or (better) refactor so that it's not a string.
- Using a template to work around private is super-ugly IMO. I see two better ways to handle this:
    1. Make it a private function of backhaul_manager. Yes, I agree that the class is already overloaded, but still...
    2. Make sRadioInfo a public class (which would imply renaming it to RadioInfo, and moving it to a separate file). This opens opportunities to add more members to RadioInfo, which is what you wanted I think. While you're at it, you'd probably want to add a class RadioList or something like that that abstracts the container.

I leave it up to you to choose which one is better.

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -100,6 +101,35 @@ namespace beerocks {
+    //          find if there is an entry with the given mac within this associated client
+
+    for (const auto &slave : slave_container) {
+        if (slave) {
+            LOG(DEBUG) << "looking into radio info for (hostap_iface): " << slave->hostap_iface;
+
+            for (const auto &client : slave->associated_clients_map) {
+                LOG(DEBUG) << "client: " << client.first;
+
+                if (client.second.find(mac) != client.second.end()) {
+                    return slave;
+                }
+            }
+        }
```
Aligned to our coding standards:

```suggestion
        if (!slave) {
            continue;
        }
        
        LOG(DEBUG) << "looking into radio info for (hostap_iface): " << slave->hostap_iface;

        for (const auto &client : slave->associated_clients_map) {
            LOG(DEBUG) << "client: " << client.first;

            if (client.second.find(mac) != client.second.end()) {
                return slave;
            }
        }
```

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -339,24 +369,25 @@ bool backhaul_manager::init()
     }
 
     if (!bus_subscribe(std::vector<ieee1905_1::eMessageType>{
-            ieee1905_1::eMessageType::TOPOLOGY_DISCOVERY_MESSAGE,
-            ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE,
-            ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_RESPONSE_MESSAGE,
+            ieee1905_1::eMessageType::ACK_MESSAGE,
             ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_RENEW_MESSAGE,
+            ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_RESPONSE_MESSAGE,
             ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE,
+            ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE,
+            ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE,
+            ieee1905_1::eMessageType::BEACON_METRICS_QUERY_MESSAGE,
             ieee1905_1::eMessageType::CHANNEL_PREFERENCE_QUERY_MESSAGE,
             ieee1905_1::eMessageType::CHANNEL_SELECTION_REQUEST_MESSAGE,
-            ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE,
-            ieee1905_1::eMessageType::CLIENT_CAPABILITY_QUERY_MESSAGE,
             ieee1905_1::eMessageType::CLIENT_ASSOCIATION_CONTROL_REQUEST_MESSAGE,
-            ieee1905_1::eMessageType::HIGHER_LAYER_DATA_MESSAGE,
+            ieee1905_1::eMessageType::CLIENT_CAPABILITY_QUERY_MESSAGE,
             ieee1905_1::eMessageType::CLIENT_STEERING_REQUEST_MESSAGE,
-            ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE,
-            ieee1905_1::eMessageType::MULTI_AP_POLICY_CONFIG_REQUEST_MESSAGE,
-            ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE,
-            ieee1905_1::eMessageType::LINK_METRIC_QUERY_MESSAGE,
             ieee1905_1::eMessageType::COMBINED_INFRASTRUCTURE_METRICS_MESSAGE,
-            ieee1905_1::eMessageType::ACK_MESSAGE,
+            ieee1905_1::eMessageType::HIGHER_LAYER_DATA_MESSAGE,
+            ieee1905_1::eMessageType::LINK_METRIC_QUERY_MESSAGE,
+            ieee1905_1::eMessageType::MULTI_AP_POLICY_CONFIG_REQUEST_MESSAGE,
+            ieee1905_1::eMessageType::TOPOLOGY_DISCOVERY_MESSAGE,
+            ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE,
+            ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE,
```
If you decide to sort the list alphabetically, please do it in a separate commit

RanRegev replied:
done.

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1650,7 +1681,8 @@ bool backhaul_manager::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_r
         // (by the Backhaul Manager), this function will return 'true'.
         // Otherwise, it should be forwarded to the slaves.
-        if (handle_1905_1_message(cmdu_rx, src_mac)) {
+        Socket *destinationAP = nullptr;
```
```suggestion
        Socket *target_slave = nullptr;
```

**BUT**, I think it's not required at all.
You can just forward the CMDU from your handler function to the correct slave and return `true` (so it won't be forwarded to all slaves here).

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2890,6 +2935,74 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
+        return false;
+    }
+
+    const sMacAddr &requiredMac = tlvBeaconMetricsQuery->associated_sta_mac();
```
We use Snake Case not Camel Case, so the name of the variable (as an example) should change to:
```suggestion
    const sMacAddr &required_mac = tlvBeaconMetricsQuery->associated_sta_mac();
```

RanRegev committed:
agent: backhaul: sort subscription list alphabetically

It is easier to read this way.
Using the code editor to sort it, it is not
done manually

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
agent: backhaul: added handling 1905 beacon metrics query

Added the code to handle 1905 beacon query with all
the required "wires" to make it happen

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
tlvf: beacon query yaml: implemented according to spec

The implemented yaml was not according to the spec.
(AP specification v.1: 17.2.27 Beacon Metrics Query TLV format)
Now it conforms to the spec.

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
Auto generated: changed beacon metrics following yaml changes

since the yaml was changed the auto-generated files were changed as
well.

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev committed:
agent: son slave: added support for ap channl list in beacon query

added the code to extract 1905 channel support list and
copy them to the vs message
not tested yet

Signed-off-by: Ran regev <ran.regev@devalore.com>

RanRegev commented:
Replaced by #1274 and #1273
